### PR TITLE
Toolbox sort

### PR DESCRIPTION
**Pull request by MysticJay**

### CODE REVIEW

File: core/code/boot.js [hunk 1]
```diff
@@ -205,6 +205,29 @@ function prepPluginsToLoad () {
...
+  ) {
+    sortedChildren.forEach(function (child) {
+      toolboxElement.removeChild(child);
+      toolboxElement.appendChild(child);
+    });
+  }
+}
+function setupToolboxSort() {
+  toolboxElement = $('#toolbox')[0];
+  window.setTimeout( ()=> window.sortToolbox(), 1000);
```
Comment (xscreach) [inline core/code/boot.js]: We don't use arrow functions in core.

```suggestion
  window.setTimeout(window.sortToolbox, 1000);
```

File: core/code/boot.js [hunk 2]
```diff
@@ -205,6 +205,29 @@ function prepPluginsToLoad () {
...
+    })
+  ) {
+    sortedChildren.forEach(function (child) {
+      toolboxElement.removeChild(child);
+      toolboxElement.appendChild(child);
+    });
+  }
+}
+function setupToolboxSort() {
+  toolboxElement = $('#toolbox')[0];
```
Comment (xscreach) [inline core/code/boot.js]: This and the variable itself should be moved into the sortToolbox function

File: core/code/utils_misc.js [hunk 3]
```diff
@@ -427,3 +427,22 @@ if (!Element.prototype.closest) {
     return null;
   };
 }
+
+var MutObserver = window.MutationObserver || window.WebKitMutationObserver;
```
Comment (xscreach) [inline core/code/utils_misc.js:431]: This is not used so it should be here.

**Review by xscreach — CHANGES_REQUESTED** [review]
Works for all mobile, button and tampermonkey (even with plugins loaded after iitc main script)

Needs code cleanup

Edit: doesn't always work with tampermonkey - large scripts might load after the sort is already done

File: core/code/boot.js (hunk 2 above)
Comment (le-jeu) [reply core/code/boot.js]:
```suggestion
```

File: core/code/boot.js [hunk 4]
```diff
@@ -205,6 +205,29 @@ function prepPluginsToLoad () {
   };
 }
 
+var toolboxElement;
+window.sortToolbox = function () {
```
Comment (le-jeu) [inline core/code/boot.js]:
```suggestion
window.sortToolbox = function () {
  var toolboxElement = document.getElementById('toolbox') ;
```

**Comment by le-jeu** [comment]
since this is about improving toolbox readability, you can also display the link by column :
image

With  `column-count: 2` (and few tweak on margin/padding):
image

Same but with 3 columns:
image

It's possible to use a grid layout as well, giving the links another direction using 
`display: grid; grid-template-column: repeat(<number of column>, 1fr)` for instance
image

File: core/code/utils_misc.js (hunk 3 above)
Comment (MysticJay) [reply core/code/utils_misc.js:431]: I just left it, as it is a nice extension to utils_misc. But maybe it is worth a different PR
@xscreach

Comment (xscreach) [reply core/code/utils_misc.js:431]: If it's not used then just remove it.
It can be simply added back when needed.

**Comment by MysticJay** [comment]
> since this is about improving toolbox readability, you can also display the link by column :
> 
> With `column-count: 2` (and few tweak on margin/padding): 
> 
> Same but with 3 columns: 
> 
> It's possible to use a grid layout as well, giving the links another direction using `display: grid; grid-template-column: repeat(<number of column>, 1fr)` for instance 

looks nice and easy to implement, but I'd rather put that to a new issue/PR.
Many users have complained in the past about  the toolbox becoming too large. Putting the links into two columns adds even more size. 

My proposal here would be a mall checkbox that switches between the old (now sorted) and the 2-columnsview (vertically sorted). Three colums seem to overdo, but maybe the checkbox could also be a 3-state-toggle between 1, 2 and 3.

File: core/code/boot.js [hunk 5]
```diff
@@ -205,6 +205,28 @@ function prepPluginsToLoad () {
...
+    })
+  ) {
+    sortedChildren.forEach(function (child) {
+      toolboxElement.removeChild(child);
+      toolboxElement.appendChild(child);
+    });
+  }
+}
+function setupToolboxSort() {
+  window.setTimeout(window.sortToolbox, 1000);
```
Comment (xscreach) [inline core/code/boot.js:227]: The timeout won't be enough to have all the links always sorted - there'll always be possibility for some plugins to load later and add unsorted links.

Imho we should keep the observer and allow plugins to override the sorting function - that way we ensure it's always sorted (100%) and we also allow extensions by plugins 🤷

Comment (MysticJay) [reply core/code/boot.js:227]: If a plugin adds to the toolbox dynamically, it should also call sortToolbxo() after that.
If we find slow plugins that have not finished initializing, we should look into this again.

**Pull request closed without merging**